In [142]:
import pandas as pd

#import simulated data csv

In [143]:
df_sim = pd.read_csv('../data/student_health_data.csv')
df_sim = df_sim.drop(columns=['Student_ID', 'Unnamed: 0'], errors='ignore').reset_index(drop=True)
df_sim.head()

Age Gender  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
0   24      M   50.663217               122.173015                 84.419860   
1   21      F   57.926042               110.778407                 75.696145   
2   22      M   59.294219               109.375673                 83.803814   
3   24      M   76.826232               125.142227                 78.091587   
4   20      M   68.342769               107.515592                 80.674937   

   Stress_Level_Biosensor  Stress_Level_Self_Report Physical_Activity  \
0                3.137350                  9.028669              High   
1                3.699078                  5.819697          Moderate   
2                6.785156                  5.892360               Low   
3                6.408509                  6.884001              High   
4                7.264719                  4.483450          Moderate   

  Sleep_Quality      Mood  Study_Hours  Project_Hours Health_Risk_Level  
0      Moderate     Happy    34.520973      16.800956          Moderate  
1          Good  Stressed    16.763846      15.791154          Moderate  
2      Moderate     Happy    44.203798      25.678437          Moderate  
3          Poor     Happy    21.776645      20.808391              High  
4          Poor     Happy     8.964999      15.194045          Moderate

#dropped 'Student_ID' and any unnamed columns.

#quick snapshot/overview of imported dataset

In [144]:
df_sim.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1000 non-null   int64  
 1   Gender                    1000 non-null   object 
 2   Heart_Rate                1000 non-null   float64
 3   Blood_Pressure_Systolic   1000 non-null   float64
 4   Blood_Pressure_Diastolic  1000 non-null   float64
 5   Stress_Level_Biosensor    1000 non-null   float64
 6   Stress_Level_Self_Report  1000 non-null   float64
 7   Physical_Activity         1000 non-null   object 
 8   Sleep_Quality             1000 non-null   object 
 9   Mood                      1000 non-null   object 
 10  Study_Hours               1000 non-null   float64
 11  Project_Hours             1000 non-null   float64
 12  Health_Risk_Level         1000 non-null   object 
dtypes: float64(7), int64(1), object(5)
memory usage: 101.7+ KB


In [145]:
print(df_sim.columns.tolist())

['Age', 'Gender', 'Heart_Rate', 'Blood_Pressure_Systolic', 'Blood_Pressure_Diastolic', 'Stress_Level_Biosensor', 'Stress_Level_Self_Report', 'Physical_Activity', 'Sleep_Quality', 'Mood', 'Study_Hours', 'Project_Hours', 'Health_Risk_Level']


    #This gives us the statistical snapshot.

In [146]:
df_sim.describe()

Age   Heart_Rate  Blood_Pressure_Systolic  \
count  1000.0000  1000.000000              1000.000000   
mean     20.9600    70.101863               119.835380   
std       2.0036     9.544663                14.604307   
min      18.0000    50.000000                90.000000   
25%      19.0000    63.347970               109.386868   
50%      21.0000    70.330916               120.051654   
75%      23.0000    76.425149               130.397799   
max      24.0000    99.650104               165.929204   

       Blood_Pressure_Diastolic  Stress_Level_Biosensor  \
count               1000.000000             1000.000000   
mean                  79.854756                5.483909   
std                    9.641699                2.596108   
min                   60.000000                1.008491   
25%                   73.115420                3.270567   
50%                   79.651738                5.368891   
75%                   86.535640                7.674233   
max                  107.659796                9.994153   

       Stress_Level_Self_Report  Study_Hours  Project_Hours  
count               1000.000000  1000.000000    1000.000000  
mean                   5.361601    30.227037      14.887368  
std                    2.550382    10.341060       4.888624  
min                    1.002186     5.000000       0.000000  
25%                    3.172409    23.262704      11.573503  
50%                    5.234889    30.432767      14.765873  
75%                    7.471178    37.245193      18.159167  
max                    9.964115    60.000000      32.720325

#Heart_Rate: min= 50, max= 100 mean= 70? 
#Study_Hours:  max= 60, mean= 30-- they did try to include high workload outliers?
 #Biosensor vs Self-reported: the machine thinks the student is slightly more stressed than the student feels?
        Stress_Level_Self_Reported: mean= 5.48
        Stress_Level_Biosensor:  mean= 5.36

#create a stats summary with .describe and .drop

In [147]:
Stats_Summary = df_sim.describe().drop(['count', 'std', '25%', '50%', '75%'])
Stats_Summary.head()

Age  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
mean  20.96   70.101863               119.835380                 79.854756   
min   18.00   50.000000                90.000000                 60.000000   
max   24.00   99.650104               165.929204                107.659796   

      Stress_Level_Biosensor  Stress_Level_Self_Report  Study_Hours  \
mean                5.483909                  5.361601    30.227037   
min                 1.008491                  1.002186     5.000000   
max                 9.994153                  9.964115    60.000000   

      Project_Hours  
mean      14.887368  
min        0.000000  
max       32.720325

#check if the simulation is biased in its population

In [148]:
df_sim['Gender'].value_counts()

Gender
M    519
F    481
Name: count, dtype: int64

The simulated data shows a balanced gender split, which may not reflect the actual demographics of specific academic majors in the MIT study.

#Heart Rate by Gender
#grouping by gender to see the average heart rate, blood pressure, and stress

In [149]:
df_sim.groupby('Gender')[['Heart_Rate', 'Blood_Pressure_Systolic', 'Blood_Pressure_Diastolic', 'Stress_Level_Biosensor', 'Stress_Level_Self_Report', 'Study_Hours', 'Project_Hours']].mean()

Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
Gender                                                                  
F        69.755754               120.617518                 79.339332   
M        70.422631               119.110508                 80.332441   

        Stress_Level_Biosensor  Stress_Level_Self_Report  Study_Hours  \
Gender                                                                  
F                     5.631104                  5.321370    29.885978   
M                     5.347492                  5.398885    30.543125   

        Project_Hours  
Gender                 
F           14.825392  
M           14.944807

The average heart rate for both genders are almost the same (both are 70.1), meaning this simulation assumes physiological uniforminity.
In the simulated environment, there appears to be an inverse relation between the heart rate and biosensor stress when comparing genders. 
    Females (F): Lower average HR (69.7), but Higher Biosensor Stress (5.63).
    Males (M): Higher average HR (70.4), but Lower Biosensor Stress (5.34).
The simulation seems to "view" genders differently, it assumes women's bodies show higher stress through the biosensor compared to men
Study and Project hours are nearly identical, meaning the stress difference isn't coming from a workload gap

My audit will investigate if this is a programed stereotype in the simulation or if the MIT real-world data reflects a similar physciological pattern.

#Using .loc to find females with very high stress

In [150]:
test_filter  = df_sim.loc[(df_sim['Gender'] == 'F') & (df_sim['Stress_Level_Biosensor'] > 7)]
test_filter.head()

Age Gender  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
6    22      F   93.097228               106.677782                 76.499815   
12   20      F   64.824663               138.396077                 84.015606   
14   21      F   75.553768               137.642627                 85.199991   
16   23      F   65.040310               123.100861                 73.176090   
31   18      F   75.838004               132.684023                 72.671071   

    Stress_Level_Biosensor  Stress_Level_Self_Report Physical_Activity  \
6                 8.415979                  1.402915          Moderate   
12                8.953156                  8.403831          Moderate   
14                7.154831                  4.546753              High   
16                7.210422                  4.601397          Moderate   
31                9.334536                  1.397310               Low   

   Sleep_Quality     Mood  Study_Hours  Project_Hours Health_Risk_Level  
6       Moderate    Happy    10.269850      23.818096          Moderate  
12          Good  Neutral    40.484849      12.992542              High  
14      Moderate  Neutral    39.763737      10.290925          Moderate  
16          Good    Happy    32.930511      18.664338          Moderate  
31      Moderate  Neutral    39.860839       7.197902          Moderate

#Using .loc to compare returned rows, the code above assumes "Average Heart Rate", but does not implicitly name it.
#The code below includes logic for "Average Heart Rate" and returns the same rows.

#Using .loc to find females with very high stress and index reset 

In [151]:
High_Stress_F_Sim  = df_sim.loc[
    (df_sim['Gender'] == 'F') &
    (df_sim['Stress_Level_Biosensor'] > 7) &
    (df_sim['Heart_Rate'] > 65) & (df_sim['Heart_Rate'] < 75)
]
High_Stress_F_Sim = High_Stress_F_Sim.reset_index(drop=True)
High_Stress_F_Sim.head()

Age Gender  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
0   23      F   65.040310               123.100861                 73.176090   
1   24      F   71.047353               103.581202                 77.942314   
2   21      F   73.431351               142.230337                 88.867324   
3   24      F   74.216849               138.199746                 76.985026   
4   19      F   74.056176               117.740911                 82.491787   

   Stress_Level_Biosensor  Stress_Level_Self_Report Physical_Activity  \
0                7.210422                  4.601397          Moderate   
1                7.674095                  4.552978               Low   
2                9.568562                  6.830533               Low   
3                7.695147                  7.952703          Moderate   
4                9.272872                  2.519502              High   

  Sleep_Quality      Mood  Study_Hours  Project_Hours Health_Risk_Level  
0          Good     Happy    32.930511      18.664338          Moderate  
1          Good   Neutral    21.270167      14.513982          Moderate  
2          Good   Neutral    12.421566      17.360057          Moderate  
3          Good     Happy    22.071364      15.438906              High  
4          Good  Stressed    37.634382       8.533486          Moderate

#Using .loc to find males with very high stress and index reset 

In [152]:
High_Stress_M_Sim  = df_sim.loc[
    (df_sim['Gender'] == 'M') &
    (df_sim['Stress_Level_Biosensor'] > 7) &
    (df_sim['Heart_Rate'] > 65 & (df_sim['Heart_Rate'] < 75))
]
High_Stress_M_Sim = High_Stress_M_Sim.reset_index(drop=True)
High_Stress_M_Sim.head()

Age Gender  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
0   20      M   68.342769               107.515592                 80.674937   
1   20      M   86.211086               112.635938                 91.951392   
2   21      M   93.573764                90.609147                 72.854530   
3   20      M   80.836097               108.817898                 88.899876   
4   24      M   76.947610                92.465136                 74.006335   

   Stress_Level_Biosensor  Stress_Level_Self_Report Physical_Activity  \
0                7.264719                  4.483450          Moderate   
1                8.208448                  5.946601          Moderate   
2                9.197436                  8.964341          Moderate   
3                9.692491                  2.173165          Moderate   
4                9.248402                  3.203199               Low   

  Sleep_Quality      Mood  Study_Hours  Project_Hours Health_Risk_Level  
0          Poor     Happy     8.964999      15.194045          Moderate  
1      Moderate     Happy    33.195541      19.744741          Moderate  
2      Moderate  Stressed    20.001047       9.941094              High  
3      Moderate   Neutral    53.112590      24.263485          Moderate  
4          Poor   Neutral    17.361192      21.960476          Moderate

#sorting each group 'Study Hours'

In [153]:
df_sim.sort_values(by='Study_Hours', ascending=False)

Age Gender  Heart_Rate  Blood_Pressure_Systolic  \
443   24      F   55.928560                90.000000   
63    18      M   74.805621               113.068560   
837   24      M   70.058106               129.669424   
61    19      F   85.316888               139.087312   
356   20      F   63.966764               113.643354   
..   ...    ...         ...                      ...   
664   21      F   53.011508               109.978940   
729   24      F   59.021376               134.377148   
122   24      M   79.505028                98.279154   
890   20      F   77.236106               114.034438   
219   18      F   71.792746               104.506962   

     Blood_Pressure_Diastolic  Stress_Level_Biosensor  \
443                 64.111815                3.508249   
63                  73.850496                9.185936   
837                 82.116004                2.518745   
61                  75.128420                5.401436   
356                 91.446063                6.380272   
..                        ...                     ...   
664                 74.025766                8.894148   
729                 83.916706                2.127261   
122                 82.293141                9.076576   
890                 79.972498                2.573032   
219                 77.077448                8.507477   

     Stress_Level_Self_Report Physical_Activity Sleep_Quality      Mood  \
443                  8.802469          Moderate          Poor  Stressed   
63                   5.587814          Moderate      Moderate     Happy   
837                  8.253731               Low          Poor     Happy   
61                   3.362577          Moderate          Good   Neutral   
356                  5.726268          Moderate      Moderate   Neutral   
..                        ...               ...           ...       ...   
664                  8.988068              High      Moderate   Neutral   
729                  8.751090          Moderate          Good     Happy   
122                  3.458437          Moderate          Good   Neutral   
890                  9.087750               Low          Good  Stressed   
219                  6.272767          Moderate          Good   Neutral   

     Study_Hours  Project_Hours Health_Risk_Level  
443    60.000000       8.509177          Moderate  
63     59.863294      20.754383          Moderate  
837    58.916587      13.646866          Moderate  
61     57.136198       8.069757          Moderate  
356    56.805089      18.615965          Moderate  
..           ...            ...               ...  
664     5.000000      23.601327              High  
729     5.000000      18.058100          Moderate  
122     5.000000       6.824012          Moderate  
890     5.000000      14.392630          Moderate  
219     5.000000      10.531732          Moderate  

[1000 rows x 13 columns]

Student has max Study_Hours (60) and has very high Self_Reported_Stress (8.80), but Heart_Rate is (55.9) very low.

In [154]:
High_Stress_F_Sim.sort_values(by='Study_Hours', ascending=False)

Age Gender  Heart_Rate  Blood_Pressure_Systolic  Blood_Pressure_Diastolic  \
34   22      F   71.476993               128.519126                 85.542247   
44   20      F   70.570131               141.625751                 62.205706   
19   20      F   73.406000               130.053298                 86.356616   
46   24      F   72.036721               137.469978                 90.499942   
37   24      F   69.929572               105.182867                 69.989378   
..  ...    ...         ...                      ...                       ...   
35   22      F   69.803765               119.751964                 97.701676   
17   18      F   72.541739               122.096655                 60.000000   
2    21      F   73.431351               142.230337                 88.867324   
28   20      F   66.469839               134.932987                 73.578053   
14   18      F   71.792746               104.506962                 77.077448   

    Stress_Level_Biosensor  Stress_Level_Self_Report Physical_Activity  \
34                7.896209                  6.882908          Moderate   
44                9.664593                  2.401765               Low   
19                8.349897                  3.593369          Moderate   
46                9.062342                  7.696445              High   
37                7.393164                  7.115967          Moderate   
..                     ...                       ...               ...   
35                8.271616                  9.603269              High   
17                7.819989                  2.545390               Low   
2                 9.568562                  6.830533               Low   
28                7.598443                  8.470294               Low   
14                8.507477                  6.272767          Moderate   

   Sleep_Quality      Mood  Study_Hours  Project_Hours Health_Risk_Level  
34          Good     Happy    55.413410       7.156937          Moderate  
44      Moderate   Neutral    55.052652      27.091956          Moderate  
19      Moderate     Happy    53.232670      13.622509          Moderate  
46          Good     Happy    47.914793      27.308475              High  
37      Moderate  Stressed    46.311151      17.910017              High  
..           ...       ...          ...            ...               ...  
35      Moderate   Neutral    14.265029      11.537094              High  
17          Good  Stressed    14.192999      12.292171          Moderate  
2           Good   Neutral    12.421566      17.360057          Moderate  
28      Moderate     Happy     7.916278      17.473523              High  
14          Good   Neutral     5.000000      10.531732          Moderate  

[61 rows x 13 columns]

In [155]:
High_Stress_M_Sim.sort_values(by='Study_Hours', ascending=False)

Age Gender  Heart_Rate  Blood_Pressure_Systolic  \
6     18      M   74.805621               113.068560   
91    24      M   90.605586               137.199359   
3     20      M   80.836097               108.817898   
79    24      M   94.648776               111.960613   
114   22      M   68.647571                97.091701   
..   ...    ...         ...                      ...   
40    21      M   64.490519               125.476881   
158   18      M   76.419305               108.582204   
0     20      M   68.342769               107.515592   
154   23      M   80.501106               117.263685   
21    24      M   79.505028                98.279154   

     Blood_Pressure_Diastolic  Stress_Level_Biosensor  \
6                   73.850496                9.185936   
91                  86.764418                7.628259   
3                   88.899876                9.692491   
79                  99.708643                9.797109   
114                 66.417883                7.270201   
..                        ...                     ...   
40                  68.481316                9.673167   
158                 76.880130                8.387641   
0                   80.674937                7.264719   
154                 65.788469                9.796944   
21                  82.293141                9.076576   

     Stress_Level_Self_Report Physical_Activity Sleep_Quality     Mood  \
6                    5.587814          Moderate      Moderate    Happy   
91                   4.152451              High      Moderate  Neutral   
3                    2.173165          Moderate      Moderate  Neutral   
79                   5.777731          Moderate          Good    Happy   
114                  9.271344          Moderate          Good  Neutral   
..                        ...               ...           ...      ...   
40                   7.133146               Low      Moderate  Neutral   
158                  8.611704               Low          Poor    Happy   
0                    4.483450          Moderate          Poor    Happy   
154                  5.560930          Moderate          Poor  Neutral   
21                   3.458437          Moderate          Good  Neutral   

     Study_Hours  Project_Hours Health_Risk_Level  
6      59.863294      20.754383          Moderate  
91     55.333390       9.811476          Moderate  
3      53.112590      24.263485          Moderate  
79     50.001077      14.372697          Moderate  
114    47.710212      16.240203              High  
..           ...            ...               ...  
40     10.793111      10.163594              High  
158     9.554437       9.339386              High  
0       8.964999      15.194045          Moderate  
154     5.736824      17.401622          Moderate  
21      5.000000       6.824012          Moderate  

[163 rows x 13 columns]

Workload vs. Health
Students studying over 55 hours a week: despite high stress level, their Health Risk level is "Moderate".

NEXT STEPS: Now that the simulated baseline is established, the next phase will involve importing the MIT Real-World Dataset.  I will apple the same cleaning techniques like removing IDs, handling categorical biases, and filtering for high-stress cohorts, to determine is the same occurences are reflected in actual student biosensor data. 